src/targets/github.ts

```diff
@@ -378,12 +378,33 @@ export class GithubTarget extends BaseTarget {
         );
       }
 
-      // XXX: This is a bit hacky as we rely on two things:
-      // 1. GitHub issuing a redirect to S3, where they store the artifacts,
-      //    or at least pass those request headers unmodified to us
-      // 2. AWS S3 using the MD5 hash of the file for its ETag cache header
-      //    when we issue a HEAD request.
-      const response = await this.github.request(`HEAD ${url}`, {
+      const remoteChecksum = await this.checksumFromUrl(url);
+      const localChecksum = this.checksumFromData(file);
+      if (localChecksum !== remoteChecksum) {
+        throw new Error(
+          `Uploaded asset checksum does not match local asset checksum for "${name} (${localChecksum} != ${remoteChecksum})`
+        );
+      }
+      uploadSpinner.succeed(`Uploaded asset "${name}".`);
+      return url;
+    } catch (e) {
+      uploadSpinner.fail(`Cannot upload asset "${name}".`);
+
+      throw e;
+    }
+  }
+
+  private async checksumFromUrl(url: string): Promise<string> {
+    // XXX: This is a bit hacky as we rely on various things:
+    // 1. GitHub issuing a redirect to AWS S3.
+    // 2. S3 using the MD5 hash of the file for its ETag cache header.
+    // 3. The file being small enough to fit in memory.
```

Comment:
Calculating MD5 hash from a streaming data is quite trivial in Node so I'd encourage you to revisit this approach in a follow up PR as it causes unnecessary memory usage. We may have rather large files.

Reply:
Yes, @iker-barriocanal and I talked extensively about this.

Trade-offs:
- Octokit's `request` method seems to buffer the full response payload, so if we wanted to work with streams we'd need to depart from using `request`. I believe we're originally using Octokit in this code path to benefit from the different rate limits given to authenticated requests -- @BYK can you confirm that?
- We don't work with streams when doing the upload. We read (with `readSync`!) full asset files from disk into memory before uploading. We use the same bytes in memory to calculate the "local checksum". So the assumption that files are small enough was implicitly already there.
- If memory usage becomes a concern, then certainly we need to revisit this, but probably then not only the fallback download step.
- We've filed issues here on GitHub and internal tracking tickets for future improvements. Perhaps none specifically about streaming -- now done in https://github.com/getsentry/craft/issues/329

Reply:
>Octokit's request method seems to buffer the full response payload, so if we wanted to work with streams we'd need to depart from using request. I believe we're originally using Octokit in this code path to benefit from the different rate limits given to authenticated requests 

Octokit takes care of a bunch of things from authentication to rate limiting to parallelization to automatic retries. So I'd say stick with it as long as it can. If solving this issue requires a departure from Octokit, delay that as long as possible and file an issue with them ahead of time so they might be able to bring this into the library itself.

>We don't work with streams when doing the upload. We read (with readSync!) full asset files from disk into memory before uploading. We use the same bytes in memory to calculate the "local checksum". So the assumption that files are small enough was implicitly already there.

Great catch. This is again due to Octokit limitations tho as it does not support streams or piping, only buffers.

Reply:
References to our future selves (?):

- https://github.com/octokit/octokit.js/issues/916#issuecomment-397742916
- https://github.com/octokit/octokit.js/issues/887#issuecomment-391448184

2018 was a fine year, right? :)